### PR TITLE
gh-157364: Fix use-after-free in TextIOWrapper during reentrant detach

### PR DESCRIPTION
**Pull request by ashm-dev**
* Issue: gh-157364

### CODE REVIEW

File: Modules/_io/textio.c [hunk 1]
```diff
@@ -753,43 +753,55 @@ buffer_access_safe(textio *self)
         return NULL;
     }
 
-    /* Returning a borrowed reference is safe since TextIOWrapper methods are
-       protected by critical sections. */
+    /* The critical section protects this borrowed reference until the caller
+       can acquire its own reference. */
     _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(self);
     return self->buffer;
```
Comment (vstinner) [inline Modules/_io/textio.c]: I would prefer to modify this function to return a strong reference, and update _io.TextIOWrapper.detach(), the last caller, to fix ref counting (replace `self->buffer = NULL;` with `Py_CLEAR(buffer);`.

File: Modules/_io/textio.c [hunk 2]
```diff
@@ -753,43 +753,55 @@ buffer_access_safe(textio *self)
...
 }
 
+static PyObject *
+buffer_acquire_safe(textio *self)
```
Comment (vstinner) [inline Modules/_io/textio.c]: It's strange to have two "safe" functions, whereas one is not really safe. Just remove it and return a strong reference in buffer_access_safe().

File: Modules/_io/textio.c (hunk 1 above)
Comment (ashm-dev) [reply Modules/_io/textio.c]: Fixed

File: Modules/_io/textio.c (hunk 2 above)
Comment (ashm-dev) [reply Modules/_io/textio.c]: Fixed

File: Lib/test/test_io/test_textio.py [hunk 3]
```diff
@@ -1633,6 +1633,33 @@ def make_text(buffer):
...
+                b[:3] = b"ab\n"
+                return 3
+
+        for method_name in ("read", "readline"):
+            with self.subTest(method_name):
+                raw = DetachOnRead()
+                wrapper = self.TextIOWrapper(
+                    self.BufferedReader(raw), encoding="utf-8")
+                method = getattr(wrapper, method_name)
+                self.assertEqual(method(), "ab\n")
```
Comment (vstinner) [inline Lib/test/test_io/test_textio.py:1661]: Can you check that *wrapper* is actually detached? Maybe get `wrapper.buffer` and expect `PyExc_ValueError("underlying buffer has been detached")`?

**Review by vstinner — APPROVED** [review]
LGTM. I just have a last request on the test.

@cmaloney: Would you mind to double check this change? You wrote the first iteration if I recall correctly.

**Comment by cmaloney** [comment]
I will need a couple more days to look at this, it seems like this is more invasive than necessary to me and makes a number of not needed for the core UAF bug report.

**Comment by cmaloney** [comment]
For gh-157364 I think the bug, and solution, is actually in Buffered I/O. This PR changes Text I/O to keep one more reference, which may also be needed. The issue though is that Buffered I/O has an internal allocation which it passes as an argument to the Raw I/O `.readinto` method. That Raw I/O call may store and use that reference for an arbitrary amount of time. As long as it is stored the Buffered I/O should not get deallocated.

That makes two pieces to fix here:
1. The Buffer Protocol object currently doesn't reference the Buffered I/O that allocates and deallocates the buffer. That means if the buffer is stored anywhere than used later you can get a use after free. No Text I/O needed.
   https://github.com/python/cpython/blob/fd0970c0ab7eb8c685ef9e5476f36e7d108c19ac/Modules/_io/bufferedio.c#L1629
2. The Buffered I/O while `buffered.read()` is executing gets de-allocated. Keeping an additional reference in the Text I/O for that case will prevent that but it also feels like "When in a method on an object the interpreter should keep that object alive".

**Comment by ashm-dev** [comment]
Thanks for looking into this, @cmaloney!

Regarding point 1: passing `(PyObject *)self` to `PyBuffer_FillInfo` at line 1629 unfortunately doesn't work. Immediately after, `PyMemoryView_FromBuffer(&buf)` explicitly clears `master.obj` (`Objects/memoryobject.c:787`):
```c
mbuf->master = *info;
mbuf->master.obj = NULL;
```
Because of this, the resulting `memoryview` still does not hold a reference to `self` (`b.obj` remains `None`). Additionally, `PyBuffer_FillInfo` acquires a new reference via `Py_XNewRef(obj)`, which `PyMemoryView_FromBuffer` drops without `Py_DECREF`, causing `self` to leak a reference on every single read. Also, for `read1()`, the underlying buffer isn't even `self->buffer`—it is transient memory from `PyBytesWriter`.

Regarding point 2: CPython method calls via `PyObject_CallMethod*` do not automatically keep `self` alive if the caller holds only a borrowed reference. Since calling `read1()` triggers arbitrary Python code (via the raw stream's `readinto`), reentrantly detaching or clearing the buffer from Python code drops the last reference. Keeping a strong reference across the call in `TextIOWrapper` is the standard pattern across CPython for guarding against reentrant deallocation here.

**Comment by cmaloney** [comment]
For 1: The `memoryview` can be stored so it needs to keep the BufferedIO whose allocation it is filling alive otherwise use after free is possible. That there is no reference is a bug. I agree the current set of calls don't give a good path there, likely other `memoryview` / Buffer Protocol methods need to be used. It looks like `PyMemoryView_FromObjectAndFlags` may be a good fit.

**Comment by ashm-dev** [comment]
@cmaloney Done in 976304399ac.
